feat: Implement Typography component with customization options

YuvrajSinghJadon opened the pull request:
Concerns ISSUE: #48

- Established the foundational Typography component to ensure consistent text styling across the application.
- Defined a scalable API supporting text variants such as h1, h2, h3, body, and caption.
- Separated component logic and styling, anticipating future integration with NativeWind for enhanced styling capabilities.
- Implemented TypeScript interfaces for strict prop typing, ensuring clarity and maintainability.
- Developed unit tests to validate component rendering and prop functionalities.
- Provided comprehensive documentation with usage examples to guide future development and integration.

## Description

This Pull Request introduces the foundational Typography component, designed to establish a consistent and scalable text styling system within the application. The component defines a clear API with support for various text variants, including headings and body text. While the current implementation focuses on structure and logic, it is designed with future integration of NativeWind in mind, allowing for enhanced styling capabilities once available. This approach ensures that the component is both forward-compatible and maintainable.

## Screenshots

To be added upon completion of styling integration.

## Tests

### Automated test cases added:

- **Render Default Variant:** Ensures the component renders correctly with default properties.
- **Apply Variant Styles:** Validates that each text variant (h1, h2, h3, body, caption) applies the intended styles.
- **Prop Functionality:** Confirms that the component correctly handles and applies provided properties.

jjalan commented on package.json:
```diff
@@ -6,6 +6,7 @@
     "android": "react-native run-android",
     "ios": "react-native run-ios",
     "lint": "eslint .",
+    "format": "prettier --write \"src/**/*.{ts,tsx,js,jsx,json,css,md}\"",
```
This should be done in a separate PR, does not tie back to the scope of the task at hand

jjalan commented on package.json:
```diff
@@ -67,7 +68,7 @@
     "jest": "^29.2.1",
     "metro-react-native-babel-preset": "0.76.5",
-    "prettier": "^2.4.1",
+    "prettier": "^2.8.8",
```
This should be done in a separate PR, does not tie back to the scope of the task at hand

jjalan commented on src/components/typography/Typography.test.tsx:
here, and other places, file name should not be case sensitive

jjalan commented on src/components/typography/Typography.test.tsx:
```diff
@@ -0,0 +1,55 @@
+// src/components/Typography/Typography.test.tsx
```
remove, here and other places

jjalan commented on src/components/typography/Typography.tsx:
```diff
@@ -0,0 +1,49 @@
+  return (
+    <Text
+      style={[
+        styles[variant], // Apply variant styles from Typography.styles.ts
```
remove comments, see [How to write better code comments](https://dev.to/adammc331/todo-write-a-better-comment-4c8c)

jjalan commented on src/components/typography/Typography.tsx:
```diff
@@ -0,0 +1,49 @@
+  children: React.ReactNode;
+}
+
+const Typography: React.FC<TypographyProps> = ({
```
fontsize, fontweight should be not accepted as props - variant should help define it.
color - believe this should be auto set from global theme.
text align - probably should come from container
children: React.ReactNode; - Not right. remove from here and rewrite `const Typography: React.FC<PropsWithChildren<TypographyProps>> = ({ `

Highly recommend to read how Tailwind, BaseWeb and others abstract Typography components and learn from it.

YuvrajSinghJadon committed:
Fix case sensitivity issue in modal.tsx

YuvrajSinghJadon committed:
Add Typography component to index exports

jjalan commented on src/components/typography/typography.tsx:
```diff
@@ -0,0 +1,36 @@
+import { styles } from './typography.styles';
+
+interface TypographyProps extends TextProps {
+  variant?: 'h1' | 'h2' | 'h3' | 'body' | 'caption';
```
Can you add in comment a reference point where this is coming from aka which library we are using as a base reference? BaseWeb / Tailwaind etc?

jjalan commented on src/components/typography/typography.tsx:
```diff
@@ -0,0 +1,36 @@
+
+interface TypographyProps extends TextProps {
+  variant?: 'h1' | 'h2' | 'h3' | 'body' | 'caption';
+  style?: object;
```
This is not expected.

By exposing `style` as props, the consumer can override it. This is seems like it would open opportunity for callers to customize the style which would lead to inconsistencies.

jjalan commented on src/components/typography/typography.tsx:
```diff
@@ -0,0 +1,36 @@
+  style?: object;
+}
+
+/**
```
Can you cross check the content of your comments here and other places, based on guidance here - [How to write better code comments](https://dev.to/adammc331/todo-write-a-better-comment-4c8c)?

jjalan commented on src/components/typography/typography.styles.ts:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Styles for different typography variants.
+ */
+export const styles = StyleSheet.create({
```
From where are we getting font size references from?

jjalan commented on src/components/typography/typography.test.tsx:
```diff
@@ -0,0 +1,43 @@
+
+import Typography from './typography';
+
+// Mock theme object (based on @react-navigation/native)
```
Can you cross check the content of your comments here and other places, based on guidance here - [How to write better code comments](https://dev.to/adammc331/todo-write-a-better-comment-4c8c)?

jjalan commented on src/components/typography/typography.styles.ts:
```diff
@@ -0,0 +1,25 @@
+import { StyleSheet } from 'react-native';
+
+/**
```
Can you cross check the content of your comments here and other places, based on guidance here - [How to write better code comments](https://dev.to/adammc331/todo-write-a-better-comment-4c8c)?

jjalan reviewed (commented):
Add screenshots or videos demonstrating the working of the implementation in the PR template

YuvrajSinghJadon committed:
feat: Introduce foundational Typography component

- Establish `Typography` component to standardize text styling.
- Define API supporting variants: `h1`, `h2`, `h3`, `body`, `caption`.
- Separate logic and styling in anticipation of NativeWind integration.
- Implement strict TypeScript interfaces for prop typing.
- Develop unit tests to validate rendering and prop handling.
- Provide documentation with usage examples.

govind-io commented on src/components/typography/typography.tsx:
```diff
@@ -0,0 +1,30 @@
+
+// Currently, it only defines the API and logic without styling
+
+type TypographyVariant = 'h1' | 'h2' | 'h3' | 'body' | 'caption';
```
Can we extend the `TypohraphyVariant` to handle the below scenarios as well?

## Heading Levels

- **"h1"** – Largest heading (e.g., page title)
- **"h2"** – Section title
- **"h3"** – Subsection title
- **"h4"** – Smaller heading
- **"h5"** – Minor heading
- **"h6"** – Smallest heading

## Text Styles

- **"body1"** – Default body text (paragraphs)
- **"body2"** – Slightly smaller body text
- **"caption"** – Small captions for images or UI elements
- **"subtitle1"** – Subtitle with medium emphasis
- **"subtitle2"** – Smaller subtitle
- **"button"** – Text used for buttons
- **"overline"** – Small uppercase text (often used for labels)

## Custom Styles (if needed)

- **"lead"** – Larger body text for emphasis
- **"muted"** – Less emphasized text (gray or faded)
- **"code"** – Monospace text for code
- **"quote"** – Styled blockquote text

These are industry standard variants for `Typography`

govind-io commented on src/components/typography/typography.tsx:
```diff
@@ -0,0 +1,30 @@
+type TypographyVariant = 'h1' | 'h2' | 'h3' | 'body' | 'caption';
+
+// TODO: Apply styles from the global theme once NativeWind is integrated
+type TypographyColor = 'primary' | 'secondary' | 'muted';
```
For colors can we add an option for `default` which inherits parent's color, and also some semantic colors such as `success`, `error`, `warning` and `info`?

YuvrajSinghJadon committed:
feat: Extend Typography component with additional variants and semantic colors

- Expanded `TypographyVariant` to support industry-standard heading levels (h1-h6), text styles (body, subtitle, button, overline), and custom styles (lead, muted, code, quote).
- Enhanced `TypographyColor` with semantic options (`success`, `error`, `warning`, `info`) and a `default` option to inherit parent color.

govind-io commented on tests/typography.test.tsx:
```diff
@@ -0,0 +1,21 @@
+  });
+
+  test('renders with different variants', () => {
+    const { getByText } = render(<Typography variant="h1">Heading</Typography>);
```
We should test, whether the correct `tag` was rendered or not based on the variant instead of just checking if the text was rendered or not. And we should do this for all possible variants, ideally by using some kind of loop.
